google-cloud-firestore/src/test/java/com/google/cloud/firestore/LocalFirestoreHelper.java

```diff
@@ -889,4 +928,84 @@ public static String autoId() {
   public static Map<String, Object> fromSingleQuotedString(String json) {
     return fromJsonString(json.replace("'", "\""));
   }
+
+  /** Map of request/response pairs to be returned when `sendRequest()` is called. */
+  static class RequestResponseStubber
+      extends LinkedHashMap<GeneratedMessageV3, ApiFuture<? extends GeneratedMessageV3>> {
+    void initializeStub(
+        ArgumentCaptor<? extends Message> argumentCaptor, FirestoreImpl firestoreMock) {
+      Stubber stubber = null;
+      for (final ApiFuture<? extends GeneratedMessageV3> response : values()) {
+        stubber = (stubber != null) ? stubber.doReturn(response) : doReturn(response);
+      }
+      if (stubber != null) {
+        stubber
+            .when(firestoreMock)
+            .sendRequest(argumentCaptor.capture(), Matchers.<UnaryCallable<Message, Message>>any());
+      }
+    }
+  }
+
+  /**
+   * Map of request/response pairs to be returned when `sendRequest()` is called.
+   *
+   * <p>Enforces that only one active request can be pending at a time.
+   */
+  static class SerialRequestStubber extends RequestResponseStubber {
+    int activeRequestCounter = 0;
+    SettableApiFuture<Void> activeRequestComplete = SettableApiFuture.create();
+    CountDownLatch latch = new CountDownLatch(1);
+
+    void markRequestComplete() {
+      activeRequestComplete.set(null);
+    }
+
+    void awaitRequest() {
+      activeRequestComplete = SettableApiFuture.create();
+      try {
+        latch.await();
+      } catch (Exception e) {
+        fail("latch.await() should not fail");
+      }
+      latch = new CountDownLatch(1);
+    }
+
+    ApiFuture<? extends GeneratedMessageV3> checkOneActiveRequestAndReturnResponse(
+        ApiFuture<? extends GeneratedMessageV3> response) {
+      ++activeRequestCounter;
+
+      // This assert is used to test that only one request is made at a time.
+      assertEquals(1, activeRequestCounter);
+      try {
+        latch.countDown();
+        activeRequestComplete.get();
+      } catch (Exception e) {
+        fail("activeRequestComplete should not fail");
+      }
+      --activeRequestCounter;
+      return response;
+    }
+
+    @Override
+    void initializeStub(
+        ArgumentCaptor<? extends Message> argumentCaptor, FirestoreImpl firestoreMock) {
+      Stubber stubber = null;
+      for (final ApiFuture<? extends GeneratedMessageV3> response : values()) {
+        Answer<ApiFuture<? extends GeneratedMessageV3>> answer =
+            new Answer<ApiFuture<? extends GeneratedMessageV3>>() {
+              @Override
+              public ApiFuture<? extends GeneratedMessageV3> answer(
+                  InvocationOnMock invocationOnMock) throws Throwable {
+                return checkOneActiveRequestAndReturnResponse(response);
```

Comment:
If you make checkOneActiveRequestAndReturnResponse() a validateRequest() method that in the base class doesn't do anything and in this class is `checkOneActiveRequestAndReturnResponse()`, then you can have a single shared implementation of `initializeStub()`.

Reply:
Great suggestion, thanks!